### PR TITLE
Add rate limiting to CL Node HTTP endpoints

### PR DESCRIPTION
**Pull request by dimroc**

### CODE REVIEW

**Comment by dimroc** [comment]
Rate limiting the Chainlink Node HTTP Server has brought to surface
more nuance than originally anticipated. Additional endpoints perform 
hashing and authentication, such as AuthenticateBridgeType using SHA256, but is
significantly faster than bcrypt.

There are four levels of granularity we could now implement.

1. Global rate limiting for all endpoints. The major con is that we want
to be especially sensitive around the /sessions endpoint.

2. Rate limiting only for /sessions endpoint. Cons is that this leaves
other authenticating endpoints like Carol's token authenticating /v2/specs
endpoint open.

3. Rate limiting endpoints not using session authentication. Con is that
now Carol and Ben (the bridge maker) is overly rate limited. i.e. 10 requests/minute
is great for POST /sessions, not great for bridges resuming runs on PATCH /v2/runs/:RunID.

4. Separate rate limits: 10/min for POST /sessions, 100/min (?) for Ben resuming bridges, etc...

I am personally leaning towards option 4, but would like to discuss more offline. Note
that having multiple rate limits would complicate configurability through Viper and env vars.

## Routes outside of Session Auth

```
POST   /sessions               --> Uses bcrypt
PATCH  /v2/runs/:RunID         --> Uses SHA256
POST   /v2/specs/:SpecID/runs  --> Uses SHA3
POST   /v2/specs               --> Uses SHA3
POST   /v2/service_agreements  --> Sealed off by dev flag, non issue
```

## Routes broken down by persona

### Carol

```
POST   /v2/specs/:SpecID/runs
POST   /v2/specs
POST   /v2/service_agreements
```

### Ben the bridge maker

```
PATCH  /v2/runs/:RunID
```

### Neil

```
GET    /debug/vars
POST   /sessions
DELETE /sessions
PATCH  /v2/user/password
GET    /v2/user/balances
POST   /v2/external_initiators
DELETE /v2/external_initiators/:AccessKey
GET    /v2/specs
GET    /v2/specs/:SpecID
DELETE /v2/specs/:SpecID
GET    /v2/runs
GET    /v2/runs/:RunID
GET    /v2/service_agreements/:SAID
GET    /v2/bridge_types
POST   /v2/bridge_types
GET    /v2/bridge_types/:BridgeName
PATCH  /v2/bridge_types/:BridgeName
DELETE /v2/bridge_types/:BridgeName
POST   /v2/withdrawals
POST   /v2/transfers
POST   /v2/keys
GET    /v2/config
GET    /v2/tx_attempts
GET    /v2/transactions
GET    /v2/transactions/:TxHash
DELETE /v2/bulk_delete_runs
```

**Comment by dimroc** [comment]
Rails' Rack Attack example configuration: https://github.com/kickstarter/rack-attack/blob/master/docs/example_configuration.md

```ruby
class Rack::Attack
  # Throttle all requests by IP (60rpm)
  throttle('req/ip', limit: 300, period: 5.minutes) do |req|
    req.ip # unless req.path.start_with?('/assets')
  end

  ### Prevent Brute-Force Login Attacks ###

  # The most common brute-force login attack is a brute-force password
  # attack where an attacker simply tries a large number of emails and
  # passwords to see if any credentials match.

  # Throttle POST requests to /login by IP address
  throttle('logins/ip', limit: 5, period: 20.seconds) do |req|
    if req.path == '/login' && req.post?
      req.ip
    end
  end

  # Throttle POST requests to /login by email param
  throttle("logins/email", limit: 5, period: 20.seconds) do |req|
    if req.path == '/login' && req.post?
      # return the email if present, nil otherwise
      req.params['email'].presence
    end
  end
end
```

**Comment by dimroc** [comment]
For this current PR, I've settled on:

1. unauthenticated POST /sessions: 5 req / 20 second. To protect against DOS by having bcrypt peg CPU. Used Rack Attack default as reference.
2. global api limit of 1000 req / minute. Our sessions are backed by SQLite, which is not as performant as other session stores like memcached. Even if requests fail, our sql store could be overburdened by a malicious DOS attacker.

Note that there are no special cases for the SHA256/SHA3 scenarios since their hash speed is fast enough to be behind the global api limit. Looking forward to discuss more offline.

**Comment by dimroc** [comment]
I have deliberately left off the ability to configure rate limits with ENV vars for the following reasons:

1. As is, this fixes the initial issue raised by hackerone without impacting existing users, therefore I lean towards YAGNI.
1. Precedence set by existing rate limiting frameworks like Rack Attack and others that hardcode sane defaults don't encourage or imply configurability without changing code, albeit their configuration is in ruby which is simpler than golang.
1. I find it rare that a Neil will want more or less than 5 login attempts per 20 seconds.
1. We have two rates, not one.

File: core/web/router.go [hunk 1]
```diff
@@ -72,18 +75,33 @@ func Router(app services.Application) *gin.Engine {
...
-		sessions.Sessions(SessionName, sessionStore),
 		secureMiddleware(config),
 	)
 
-	metricRoutes(app, engine)
-	sessionRoutes(app, engine)
-	v2Routes(app, engine)
+	api := engine.Group(
+		"/",
+		rateLimiter(1*time.Minute, 1000),
```
Comment (rupurt) [inline core/web/router.go]: 👍

File: core/web/router.go [hunk 2]
```diff
@@ -181,30 +199,31 @@ func tokenAuthRequired(store *store.Store) gin.HandlerFunc {
...
-func metricRoutes(app services.Application, engine *gin.Engine) {
-	auth := engine.Group("/", sessionAuthRequired(app.GetStore()))
+func metricRoutes(app services.Application, r *gin.RouterGroup) {
+	auth := r.Group("/", sessionAuthRequired(app.GetStore()))
 	auth.GET("/debug/vars", expvar.Handler())
 }
 
-func sessionRoutes(app services.Application, engine *gin.Engine) {
+func sessionRoutes(app services.Application, r *gin.RouterGroup) {
+	unauth := r.Group("/", rateLimiter(20*time.Second, 5))
```
Comment (rupurt) [inline core/web/router.go]: 👍 This is much simpler than I had imagined. Nice one 💯

**Pull request merged**